src/LazyJwtCredentialsManager.ts

```diff
@@ -0,0 +1,65 @@
+import { CredentialsManager } from './CredentialsManager';
+import { JwtCredentials } from './JwtCredentialsManager';
+
+const SECONDS_TO_MS = 1000;
+
+export default abstract class LazyJwtCredentialsManager<
+  TCredentials extends JwtCredentials
+> implements CredentialsManager<TCredentials> {
+  private token: string | null | undefined;
+
+  private credentialsPromise: Promise<TCredentials | null | undefined> | null;
+
+  constructor() {
+    this.token = null;
+    this.credentialsPromise = null;
+  }
+
+  private isExpired(credentials: TCredentials) {
+    return Date.now() >= credentials.exp * SECONDS_TO_MS;
+  }
+
+  async getCredentials(): Promise<TCredentials | null | undefined> {
+    const credentials = await this.credentialsPromise;
+
+    if (credentials && this.isExpired(credentials)) {
+      return this.updateCredentials();
+    }
+
+    return credentials;
+  }
+
+  abstract getCredentialsFromAuthorization(
+    authorization: string,
+  ):
+    | (TCredentials | null | undefined)
+    | Promise<TCredentials | null | undefined>;
+
+  async authenticate(token: string) {
+    this.token = token;
+    await this.updateCredentials();
```

Comment:
do you need to call it here? I guess doesn't hurt?

Reply:
you need it to pull the credentials from the token now that it exists